libqtile/backend/wayland/qw/output.c

```diff
@@ -156,3 +157,159 @@ void qw_server_output_new(struct qw_server *server, struct wlr_output *wlr_outpu
         wlr_output_layout_add_auto(server->output_layout, wlr_output);
     wlr_scene_output_layout_add_output(server->scene_layout, l_output, output->scene);
 }
+
+void qw_output_background_destroy(struct qw_output *output) {
+    // Destroy the color scene_rect if it exists
+    struct wlr_scene_rect *color_rect = output->background.color_rect;
+    if (color_rect != NULL) {
+        wlr_scene_node_destroy(&color_rect->node);
+        output->background.color_rect = NULL;
+    }
+
+    // Destroy wallpaper resources if wallpaper exists
+    if (output->background.wallpaper != NULL) {
+        struct wlr_scene_buffer *buffer = output->background.wallpaper->buffer;
+        if (buffer != NULL) {
+            wlr_scene_node_destroy(&buffer->node);
+            output->background.wallpaper->buffer = NULL;
+        }
+
+        cairo_surface_t *surface = output->background.wallpaper->surface;
+        if (surface != NULL) {
+            cairo_surface_destroy(surface);
+            output->background.wallpaper->surface = NULL;
+        }
+
+        free(output->background.wallpaper);
+        output->background.wallpaper = NULL;
+    }
+}
+
+
+void qw_output_paint_wallpaper(struct qw_output *output, cairo_surface_t *source, int mode) {
```

Comment:
I think it would be better to type the modes with a `enum` rather than a `int`, so that the switch ensure a 1:1 mapping (`-Wswitch-enum`)